### ISSUE TITLE
Object relations

### ISSUE THREAD

**Issue by ProgrammingEvangelist** [jonwagner/Insight.Database#182]
There seems to be a fundamental issue in that Guids seem to be cleared out when dealing with relationships such as parent/child.

When a child record is returned it has a Guid property signifying its parent, once the correct hierarchy is created - the property seems to be cleared out or assigned to default(Guid).  Why does this happen?  Is this a bug or by design?

**Comment by jonwagner** [comment]
I can't think of why there would be an issue specifically related to Guids. Can you provide an example of code that isn't behaving as expected?

The rules for auto-assembly are relatively straightforward, but I may need to do better docs or more examples.

**Comment by kevinsea** [comment]
I saw this behavior as well while first trying out the various ways to
indicate parent-child.  The data would come back mapped to a parent but
parent ID would not be properly set on the child.  It seemed to be
happening while doing what I would call "last chance" mapping (Jon it would
be the "auto group by" with the Guardian to do positional mapping).

I then focused on improving the mapping logic by using the parent to help
resolve the key and I forgot I saw this issue.  Let me see if I can
reproduce it and turn that into a failing test.

In the meantime, you can probably get you code to work one of two ways:
1) Add [RecordId]  and [ParentRecordId] to your objects to make the mapping
explicit.

2) add a groupby, eg:

 db.Query("GroupQuote_LoadFull_p", new { GroupQuote_id = 123}
                   , Query.Returns(Some<GroupQuote>.Records)

.ThenChildren(Some<GroupQuoteItem>.Records.GroupBy(GroupQuoteItem =>
GroupQuoteItem.GroupQuote_id)));

Either approach gives Insight much more information to do the mapping

**Comment by kevinsea** [comment]
I created a pull request with a failing test.

**Comment by ProgrammingEvangelist** [comment]
I apologize for not getting back to this sooner...  Thank you for looking into it Kevin!

Kevin you are correct and have outlined what it is that I'm now experiencing.  While our application does utilize the `[RecordIdAttribute]`, we do not specify `[ParentRecordIdAttribute]`.  Nor are we performing any grouping via the `GroupBy` function.  We simply have some standard parent/child relationships and after a call similar to:

    db.Query("spu_ReadSomeHierarchy",
                    new { ID = 7 },
                    Query.Returns(Some<Parent>.Records)
                              .ThenChildren(Some<Child>.Records, 
                                                      p => p.ID, 
                                                      (p, list) => p.Children = list);

While the Child (within the `List<Child> Children` property) instances have a `ParentID` property, and the column is corrected returned from the T-SQL stored proc, Insight doesn't correct assign the value.  I look forward to hearing back from you.

Cheers

**Comment by ProgrammingEvangelist** [comment]
The child's parent property identifier namely, `OwnerGuid` is not marked with any attributes.  However, in addition to the `[RecordId]` I added the `[ParentRecordId]` attribute to the parent's identifying property named `Key`, such that it is now as follows: `[RecordId, ParentRecordId] Guid Key { get; set; }`.

It does not fix the issue and the child still doesn't have its parent property assigned, i.e.; the OwnerGuid is still `Guid.Empty`.

**Comment by kevinsea** [comment]
The parentrecordid needs to on the child class, it is saying 'this is my
parent's id'

**Comment by ProgrammingEvangelist** [comment]
That worked, thank you!

**Comment by kevinsea** [comment]
Sorry, @jonwagner I missed your last comment somehow:

> This behavior goes back to "a time of less flexible grouping," where all of the grouping fields were external to the objects. So, it's behaving as it's supposed to, but I think expectations have changed. Given that Insight now supports more grouping options, I think it might be time to change the last resort behavior...

TL;DR I don't think supporting what I expected to happen is worth any effort or additional complexity.

The interesting thing is how my test and David's code behaves. It puts the children under the parent but other times Insight can't do the mapping, the children just don't show up...that's what confused me, it seemed like I finally had some success and a bug.  

I don't think there is a need to extend the functionality, there seem to be enough great ways to specify the parent-child relationships:

- [RecordId] / [ParentRecordId]  Attributes 
- Group By
- Or you can tell it with hints on the query
- Or more recently by matching the the ParentID fields if you want a fluent option
- For interfaces, you can add hints on the Recordset attribute:

I think just spelling out how the last chance option works is all that needed.  I also think the GroupBy option is really great -  its the perfect 'works out of the box' option.  Its not mentioned on the /Specifying-Result-Structures page,  Shouldn't this be the lead in example? - I don't think this is any more complicated than the current example which would instead look something like this:

```C#
    var results = connection.Query("GetBeerAndPossibleGlasses", Parameters.Empty,
        Query.Returns(Some<Beer>.Records)
            .ThenChildren(Some<Glasses>.Records)
            .GroupBy(child =>child.BeerId));
```
Given all of the options, I think you could make the case for dropping 'last chance' option in a v6...or maybe just deprecating in the documentation?
 
Last it might be interesting to a have '/All-The-Ways-You-Can-Do-Parent-Child-Mapping'  or '/Advanced-Parent-Child-Mapping'  page?  It would allow the Specifying-Result-Structures page to be more concise and the advanced page could show more options.

Best,

Kevin

**Comment by jonwagner** [comment]
Definitely should add groupby to that wiki page, and probably add more examples. Actually, I'd love to find someone willing to do a once-over on all of the docs. I tried to write up everything as it was developed, and now it's time to take a look at it.

As for the default assembly...I try to make everything "just magically work the way you would expect." In most cases, it does, but now that I've added the additional grouping features, there's definitely an issue where if you forget to add a hint, Insight will try to figure it out for you silently, and then you're there scratching your head.

Usually when I get to this point with a feature, if I noodle on it enough, I find a better way to do it, like a Grand Unified Theory of Everything. So I'm going to reopen this so my brain keeps thinking about this on a background thread...
